src/MPoly.jl

```diff
@@ -1271,69 +1271,67 @@ end
 ###############################################################################
 
 @doc Markdown.doc"""
-    polynomial_ring(R::AbstractAlgebra.Ring, s::Vector{T}; cached::Bool = true, ordering::Symbol = :lex) where T <: Union{String, Char, Symbol}
+    polynomial_ring(R::Ring, s::Vector{T}; cached::Bool = true, ordering::Symbol = :lex) where T <: Union{Symbol, AbstractString, Char}
 
-Given a base ring `R` and an array of strings `s` specifying how the
+Given a base ring `R` and an array of symbols/strings `s` specifying how the
 generators (variables) should be printed, return a tuple `T, (x1, x2, ...)`
 representing the new polynomial ring $T = R[x1, x2, ...]$ and the generators
-$x1, x2, ...$ of the polynomial ring. By default the parent object `T` will
-depend only on `R` and `x1, x2, ...` and will be cached. Setting the optional
-argument `cached` to `false` will prevent the parent object `T` from being
-cached. `S` is a symbol corresponding to the ordering of the polynomial and
-can be one of `:lex`, `:deglex` or `:degrevlex`.
+$x1, x2, ...$ of the polynomial ring.
+
+By default the parent object `T` depends only on `R` and `x1, x2, ...` and will be cached.
+Setting the optional argument `cached` to `false` will prevent the parent object `T` from being cached.
+
+The `ordering` of the polynomial can be one of `:lex`, `:deglex` or `:degrevlex`.
 """
-function polynomial_ring(R::AbstractAlgebra.Ring, s::Vector{String};
-                                   cached::Bool = true, ordering::Symbol = :lex)
-   return polynomial_ring(R, [Symbol(v) for v in s];
-                                   cached=cached, ordering=ordering)
-end
+polynomial_ring(R::Ring, s::Union{Tuple{Vararg{T}}, AbstractVector{T}}; kw...) where
+      T<:Union{Symbol, AbstractString, Char} =
+   polynomial_ring(R, Symbol[Symbol(x) for x in s]; kw...)
 
-function polynomial_ring(R::AbstractAlgebra.Ring, s::Vector{Char};
-                                   cached::Bool = true, ordering::Symbol = :lex)
-   return polynomial_ring(R, [Symbol(v) for v in s];
-                                   cached=cached, ordering=ordering)
+function polynomial_ring(R::Ring, s::Vector{Symbol}; kw...)
+   S = polynomial_ring_only(R, s; kw...)
+   (S, gens(S))
 end
 
-function polynomial_ring(R::AbstractAlgebra.Ring, s::Vector{Symbol};
-                                   cached::Bool = true, ordering::Symbol = :lex)
-   return Generic.polynomial_ring(R, s; cached=cached, ordering=ordering)
-end
+@doc md"""
+    polynomial_ring_only(R::Ring, s::Vector{Symbol}; ordering::Symbol=:lex, cached::Bool=true)
+
+Like [`polynomial_ring(R::Ring, s::Vector{Symbol})`](@ref) but return only the
+multivariate polynomial ring.
+"""
+polynomial_ring_only(R::T, s::Vector{Symbol}; ordering::Symbol=:lex, cached::Bool=true) where T<:Ring =
+   mpolynomial_ring_type(T)(R, s, ordering, cached)
+
+@doc md"""
+    mpolynomial_ring_type(::Type{T}) where T<:NCRing
```

Comment:
Note: we already have `mpoly_type`, so this function seems equivalent to `parent_type(mpoly_type(elem_type(T)))`, so maybe we don't need it?

Reply:
Nice. Actually, we even have `mpoly_ring_type` in [mpoly_types.jl](https://github.com/oscar-system/Oscar.jl/blob/master/src/Rings/mpoly_types.jl) in Oscar. So, I will move `mpoly_type` to AbstractAlgebra and Nemo and provide `mpoly_ring_type` via `parent_type` and `elem_type`.